### PR TITLE
feat: rewrite spore-sdk in ccc

### PR DESCRIPTION
**Pull request by ashuralyk**
# Description

1. rewrite `spore-sdk` in ccc, only contain features for spore and cluster
2. add `type-burn` and `lock-proxy` as known scripts
3. add DOB related tools
4. add examples

### CODE REVIEW

File: packages/spore/src/advanced.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+      type: "SighashAll",
+      value: {
+        seal: "0x",
+        message: {
+          actions,
+        },
+      },
+    }),
+  );
+  let txSkeleton = ccc.Transaction.from(tx);
```
Comment (Hanssen0) [inline packages/spore/src/advanced.ts:198]: `const` should works here.

File: packages/spore/src/advanced.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
+import { ccc } from "@ckb-ccc/core";
+import { bytes, UnpackResult } from "@ckb-lumos/codec";
+import { Action, ActionVec, SporeAction, WitnessLayout } from "./codec";
+import { buildProtoclScript, SporeScriptInfo } from "./predefined";
+
+const SCRIPT_INFO_HASH = ccc.hashCkb(ccc.bytesFrom("hello, cobuild", "utf8"));
+
+export async function balanceAndSignTransaction(
```
Comment (Hanssen0) [inline packages/spore/src/advanced.ts:8]: This function seems no needed anymore. Consider deleting it?

File: packages/spore/src/api/cluster.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  sporeScriptInfo?: SporeScriptInfo;
+  tx?: ccc.TransactionLike;
+}): Promise<{
+  transaction: ccc.Transaction,
+  actions: UnpackResult<typeof ActionVec>,
+}> {
+  const { signer, clusterId, tx, clusterOwner, sporeScriptInfo } = params;
+
+  // prepare transaction
+  let txSkeleton = ccc.Transaction.from(tx ?? {});
```
Comment (Hanssen0) [inline packages/spore/src/api/cluster.ts:110]: `const` should works here.

File: packages/spore/src/api/spore.ts [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        case "clusterCell": {
+          const clusterInput = ccc.CellInput.from({
+            previousOutput: clusterCell.outPoint,
+            ...clusterCell,
+          });
+          txSkeleton.inputs.push(clusterInput);
+          txSkeleton.witnesses.push("0x");
+          txSkeleton.addOutput(clusterCell.cellOutput, clusterCell.outputData);
+          txSkeleton.addCellDeps(clusterCelldep);
+          txSkeleton.addCellDepInfos(
```
Comment (Hanssen0) [inline packages/spore/src/api/spore.ts:133]: should have a await here?

File: packages/spore/src/api/cluster.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+ * @param clusterOwner the new owner of the Cluster cell
+ * @param sporeScriptInfo the script info of Spore cell, if not provided, the default script info will be used
+ * @param tx the transaction skeleton, if not provided, a new one will be created
+ * @returns
+ *  - **transaction**: a new transaction that contains transferred Cluster cell
+ *  - **actions**: cobuild actions that can be used to generate cobuild proof
+ */
+export async function transferClusterCell(params: {
+  signer: ccc.Signer;
+  clusterId: ccc.Hex;
```
Comment (Hanssen0) [inline packages/spore/src/api/cluster.ts:99]: Please use `ccc.HexLike` here and below and use `ccc.hexFrom` to convert.

File: packages/spore/src/codec/helper.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { Hex } from "@ckb-ccc/core";
+import { bytes, BytesLike } from "@ckb-lumos/codec";
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+export function bytifyRawString(text: string): Uint8Array {
```
Comment (Hanssen0) [inline packages/spore/src/codec/helper.ts:7]: consider using `ccc.bytesFrom(text, "utf8")`?

File: packages/spore/src/codec/helper.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  return encoder.encode(text);
+}
+
+export function bufferToRawString(
```
Comment (Hanssen0) [inline packages/spore/src/codec/helper.ts:11]: and `ccc.bytesTo` here.

File: packages/spore/src/codec/helper.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  source: BytesLike,
+  options?: TextDecodeOptions,
+): string {
+  const buffer = bytes.bytify(source);
+  return decoder.decode(buffer, options);
+}
+
+export function hexify(raw: string | undefined): Hex | undefined {
```
Comment (Hanssen0) [inline packages/spore/src/codec/helper.ts:19]: `ccc.hexFrom`

File: packages/spore/src/dob/api/decode.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  const sporeId = sporeOutput.type?.args;
+  if (sporeId === undefined) {
+    throw new Error("Invalid spore cell: missing spore id");
+  }
+  return decodeDobBySporeId(hexFrom(sporeId), dobServerUrl);
+}
+
+export async function decodeDobBySporeOutpoint(
+  client: Client,
+  sporeOutpoint: OutPoint,
```
Comment (Hanssen0) [inline packages/spore/src/dob/api/decode.ts:49]: `ccc.OutPointLike` should be enough.

File: packages/spore/src/dob/api/decode.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  dobServerUrl: string,
+): Promise<RenderOutput> {
+  const liveCell = await client.getCell(sporeOutpoint);
+  if (!liveCell) {
+    throw new Error("Invalid spore outpoint: missing spore cell");
+  }
+  return decodeDobBySporeCell(liveCell!.cellOutput, dobServerUrl);
```
Comment (Hanssen0) [inline packages/spore/src/dob/api/decode.ts:56]: the null ignoring is not needed here.

File: packages/spore/src/dob/config/predefined.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+const TESTNET_DECODERS = {
+  DOB0: {
+    code_hash: {
+      type: "code_hash",
+      hash: "0x13cac78ad8482202f18f9df4ea707611c35f994375fa03ae79121312dda9925c",
+    },
+    type_id: null,
+    type_script: null,
+  },
+  DOB1: {
```
Comment (Hanssen0) [inline packages/spore/src/dob/config/predefined.ts:12]: We can choose one for the devs. Experienced devs can handle these configs by themselves.

File: packages/spore/src/dob/config/predefined.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  locationType?: "code_hash" | "type_id" | "type_script",
+): Decoder {
+  locationType = locationType || "code_hash";
+  const decoder = TESTNET_DECODERS[dobVersion][locationType];
+  if (decoder === null) {
+    throw new Error(
+      `Invalid decoder: ${dobVersion} does not support ${locationType}`,
+    );
+  }
+  return decoder as Decoder;
```
Comment (Hanssen0) [inline packages/spore/src/dob/config/predefined.ts:74]: Can we add this to the global configs' type?

File: packages/spore/src/dob/config/predefined.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    },
+  },
+};
+
+export function getTestnetDecoder(
+  dobVersion: "DOB0" | "DOB1",
+  locationType?: "code_hash" | "type_id" | "type_script",
+): Decoder {
+  locationType = locationType || "code_hash";
+  const decoder = TESTNET_DECODERS[dobVersion][locationType];
```
Comment (Hanssen0) [inline packages/spore/src/dob/config/predefined.ts:68]: A function `getDecoder` accept `client` and optional `decoders?` as arg, use the `decoder` if specified and choose based on `client.addressPrefix` would be better.

File: packages/spore/src/helper.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    throw new Error("No live cell found in address: " + address);
+  }
+  let txSkeleton = ccc.Transaction.from(tx);
+  txSkeleton.inputs.push(
+    ccc.CellInput.from({
+      previousOutput: liveCell.outPoint,
+      ...liveCell,
+    }),
+  );
+  txSkeleton.witnesses.push("0x");
```
Comment (Hanssen0) [inline packages/spore/src/helper.ts:33]: Is it possible to use `Signer.prepareTransaction` after collecting all inputs? Extra witnesses don't look necessary.

File: packages/spore/src/helper.ts [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+  return txSkeleton;
+}
+
+export function computeTypeId(
+  tx: ccc.TransactionLike,
+  outputIndex: number,
+): ccc.Hex {
+  const firstInput = tx.inputs ? tx.inputs[0] : void 0;
```
Comment (Hanssen0) [inline packages/spore/src/helper.ts:41]: `tx.inputs` always exist.

File: packages/spore/src/predefined.ts [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+  procotol: SporeScript,
+  args: ccc.HexLike,
+  scriptInfo?: SporeScriptInfo,
+): ccc.Script {
+  if (scriptInfo) {
+    return ccc.Script.from({
+      args,
+      ...scriptInfo[procotol],
+    });
+  } else {
```
Comment (Hanssen0) [inline packages/spore/src/predefined.ts:96]: `else` here is not needed because we returned before.

File: packages/spore/vitest.config.mts [hunk 17]
```diff
@@ -0,0 +1,17 @@
+import { defineConfig } from "vitest/config";
```
Comment (Hanssen0) [inline packages/spore/vitest.config.mts:1]: The test framework for CCC is jest. CCC is also built for backend developers.

File: pnpm-lock.yaml [hunk 18]
```diff
@@ -1,4 +1,4 @@
-lockfileVersion: '9.0'
+lockfileVersion: '6.0'
```
Comment (Hanssen0) [inline pnpm-lock.yaml:1]: Please upgrade your pnpm version.

File: packages/spore/src/advanced.ts [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+  signer: ccc.Signer,
+  tx: ccc.TransactionLike,
+  cobuildActions: UnpackResult<typeof ActionVec>,
+  feeRate: number = 1000,
```
[the remark this answers is not in the document]
Comment (ashuralyk) [reply packages/spore/src/advanced.ts:12]: ok, I'll leave comment to tell users 1000 is the minimum fee rate

File: packages/spore/src/codec/helper.ts (hunk 7 above)
Comment (ashuralyk) [reply packages/spore/src/codec/helper.ts:11]: ccc.bytesTo cannot accept a BytesLike from Lumos, which is used in Lumos' molecule package

File: packages/spore/src/helper.ts (hunk 15 above)
Comment (ashuralyk) [reply packages/spore/src/helper.ts:41]: inputs is nullable in `TransactionLike`

**Comment by ashuralyk** [comment]
@Hanssen0 I've solved all reported issues in comment, new commit has been committed, please repeat your review

**Comment by Hanssen0** [comment]
Merged in #74

**Pull request closed without merging**